File: day9/day9.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

void part1(vector<vector<int>> heightMap);
void part2(vector<vector<int>> heightMap);
vector<vector<int>> readInput();
vector<int> getLowPoints(vector<vector<int>> &heightMap);
int getSumOfRiskLevels(vector<int> &lowPoints);
int floodFill(int x, int y, vector<vector<int>> &heightMap, int size);
vector<int> getBasinSizes(vector<vector<int>> &heightMap);
int getPart2Answer(vector<int> &basinSizes);


int main() {
    vector<vector<int>> heightMap = readInput();
    part1(heightMap);
    part2(heightMap);
    return 0;
}


void part1(vector<vector<int>> heightMap) {
    vector<int> lowPoints = getLowPoints(heightMap);
    int sumOfRiskLevels = getSumOfRiskLevels(lowPoints);
    cout << "Part 1: " << sumOfRiskLevels << endl;
}

void part2(vector<vector<int>> heightMap) {
    vector<int> basinSizes = getBasinSizes(heightMap);
    int answer = getPart2Answer(basinSizes);
    cout << "Part 2: " << answer << endl;
}

vector<vector<int>> readInput() {
    vector<vector<int>> heightMap;
    string line;
    bool isFirstLine = true;
    while (cin >> line) {
        if (isFirstLine) {
            vector<int> firstRow(line.size() + 2, 10);
            heightMap.push_back(firstRow);
            isFirstLine = false;
        }
        vector<int> row;
        row.push_back(10);
        for (char number : line) {
            int digit = number - '0';
            row.push_back(digit);
        }
        row.push_back(10);
        heightMap.push_back(row);
    }
    vector<int> lastRow(line.size() + 2, 10);
    heightMap.push_back(lastRow);
    return heightMap;
}

vector<int> getLowPoints(vector<vector<int>> &heightMap) {
    vector<int> lowPoints;
    for (int i = 1; i < heightMap.size() - 1; i++) {
        for (int j = 1; j < heightMap[0].size() - 1; j++) {
            int value = heightMap[i][j];
            if (value < heightMap[i][j - 1] and value < heightMap[i][j + 1] and value < heightMap[i - 1][j] and value < heightMap[i + 1][j]) {
                lowPoints.push_back(value);
            }
        }
    }
    return lowPoints;
}

int getSumOfRiskLevels(vector<int> &lowPoints) {
    int totalRiskLevels = 0;
    for (auto point : lowPoints) {
        totalRiskLevels += (point + 1);
    }
    return totalRiskLevels;
}

int floodFill(int x, int y, vector<vector<int>> &heightMap, int size) {
    if (heightMap[x][y] > 8) {
        return size;
    }
    heightMap[x][y] = 9;

    int size1 = floodFill(x + 1, y, heightMap, size + 1);
    int size2 = floodFill(x - 1, y, heightMap, size1);
    int size3 = floodFill(x, y + 1, heightMap, size2);
    int size4 = floodFill(x, y - 1, heightMap, size3);

    return size4;
}

vector<int> getBasinSizes(vector<vector<int>> &heightMap) {
    vector<int> basinSizes;
    for (int i = 1; i < heightMap.size() - 1; i++) {
        for (int j = 1; j < heightMap[0].size() - 1; j++) {
            if (heightMap[i][j] < 9) {
                int basinSize = floodFill(i, j, heightMap, 0);
                basinSizes.push_back(basinSize);
            }
        }
    }
    return basinSizes;
}

int getPart2Answer(vector<int> &basinSizes) {
    sort(basinSizes.begin(), basinSizes.end());
    int answer = 1;
    for (int i = basinSizes.size() - 3; i < basinSizes.size(); i++) {
        answer *= basinSizes[i];
    }
    return answer;
}

File: day9/day9.py
def read_input():
    file = open('input.in', 'r')
    height_map = [[10] + [int(i) for i in line.strip()] + [10] for line in file.readlines()]
    height_map = [[10] * len(height_map[0])] + height_map + [[10] * len(height_map[0])]
    return height_map


def get_low_points(height_map):
    low_points = []
    for i in range(1, len(height_map) - 1):
        for j in range(1, len(height_map[0]) - 1):
            value = height_map[i][j]
            if value < height_map[i][j-1] and value < height_map[i][j+1]\
                    and value < height_map[i-1][j] and value < height_map[i+1][j]:
                low_points.append(value)
    return low_points


def get_sum_of_risk_levels(low_points):
    return sum([x + 1 for x in low_points])


def flood_fill(x, y, height_map, size):
    if height_map[x][y] > 8:
        return size
    height_map[x][y] = 9

    size1 = flood_fill(x + 1, y, height_map, size + 1)
    size2 = flood_fill(x - 1, y, height_map, size1)
    size3 = flood_fill(x, y - 1, height_map, size2)
    size4 = flood_fill(x, y + 1, height_map, size3)

    return size4


def get_basin_sizes(height_map):
    sizes = []
    for i in range(1, len(height_map) - 1):
        for j in range(1, len(height_map[0]) - 1):
            if height_map[i][j] < 9:
                size = flood_fill(i, j, height_map, 0)
                sizes.append(size)
    return sizes


def get_answer(sizes):
    sizes.sort()
    largest_3 = sizes[-3:]
    answer = 1
    for size in largest_3:
        answer = answer * size
    return answer


def part1():
    height_map = read_input()
    low_points = get_low_points(height_map)
    sum_of_risk_levels = get_sum_of_risk_levels(low_points)
    print(f"Part 1: {sum_of_risk_levels}")


def part2():
    height_map = read_input()
    sizes = get_basin_sizes(height_map)
    answer = get_answer(sizes)
    print(f"Part 2: {answer}")


if __name__ == "__main__":
    part1()
    part2()
